covid19-dashboard_NEW/calculate_data.py

```diff
@@ -0,0 +1,173 @@
+    date_to_val_list = date_to_val_list[::step_size]
+
+    # Remove dates before March 1st.
+    date_to_val_list = [
+        (date, value)
+        for date, value in date_to_val_list
+        if date >= start_date
+    ]
+
+    # Remove negative values from the list.
+    date_to_val_list = [
+        (date, 0) if value < 0 else (date, value)
+        for date, value in date_to_val_list
+    ]
+
+    # Because the latest date might have been skipped by the step_size.
+    # Check to see if it's in our dataset, otherwise append it.
+    # We need the latest date to be present.
+    if date_to_val_list[-1][0] != latest_date:
+        date_to_val_list.append((latest_date, date_to_val[latest_date]))
+
+    # Convert the list of (date, value) back to a dictionary.
+    output = dict(date_to_val_list)
+
+    return output
+
+
+def calculate_data(data_type: str, data: dict, population: int) -> dict:
+    """
+    Given a dataset for a region and its population,
+    perform different types of calculations on every date.
+    :param data_type: type of data, "cases" or "deaths".
+    :param data: a dictionary containing date->int.
+    :param population: geoId -> population.
+    :return: a dictionary with the 7-week moving average + other calculations.
+    """
+    # For example, 1-week moving average.
+    chunk_size: int = 7
+
+    # Instantiate the variables to {} or None depending on data type.
+    week_ave: dict = {}
+    week_ave_value: Union[float, None] = None
+    per_capita: dict = {}
+    per_capita_value: Union[float, None] = None
+    pct_increase_value: Union[float, None] = None
+
+    # If there are less elements than our chunk_size.
+    # We can't perform any calculations.
+    if len(data.keys()) > chunk_size:
+        # The latest date is the largest date from the list of dates.
+        latest_date: str = max(list(data.keys()))
+
+        # Calculate the time-series moving average.
+        week_ave: dict = _moving_average(data, chunk=chunk_size)
+        # Get the value for the current date.
+        week_ave_value: int = week_ave[latest_date]
+
+        # If population is larger than 0, perform the division on dataset.
+        if population > 0:
+            per_capita: dict = {
+                key: (week_ave[key] / population) * 1_000_000
+                for key in week_ave
+            }
+            per_capita_value: float = per_capita[latest_date]
+
+        # Calculate the latest_date - 7 days.
+        date_1_week_ago: datetime = \
+            datetime.fromisoformat(latest_date) - timedelta(days=7)
+        date_1_week_ago: str = date_1_week_ago.strftime('%Y-%m-%d')
+
+        # If the date 1 week ago is present in the dataset.
+        if date_1_week_ago in week_ave:
+            # Get the value for that date.
+            week_ave_1w_value: int = week_ave[date_1_week_ago]
+            # If the value is larger than 0, perform pct change.
+            if week_ave_1w_value > 0:
```

Comment:
can we better handle these negative values? if they are corrective, can we perform the correction and ensure positive numbers throughout?

Reply:
For example, in this time-series:

Day 1: 10 new deaths
Day 2: -5 new deaths
Day 3: 15 new deaths

-5 usually means they corrected the dataset.
Should we replace Day 2 value with a 0, instead?